File: scripts/build_submission_zip.py
#!/usr/bin/env python3
"""Build a deterministic skills-only ZIP for the OpenAI submission portal."""

from __future__ import annotations

import argparse
import hashlib
import json
import stat
import tempfile
import zipfile
from pathlib import Path

from validate_submission import validate_repository


INCLUDED_PATHS = (
    Path(".codex-plugin"),
    Path("assets"),
    Path("skills"),
    Path("LICENSE"),
    Path("README.md"),
    Path("package.json"),
)
ZIP_TIMESTAMP = (1980, 1, 1, 0, 0, 0)


def collect_files(root: Path) -> list[Path]:
    files: list[Path] = []
    for relative in INCLUDED_PATHS:
        path = root / relative
        if path.is_symlink():
            raise ValueError(f"Submission archives cannot include symlinks: {relative}")
        if path.is_file():
            files.append(path)
            continue
        if not path.is_dir():
            raise ValueError(f"Required submission path is missing: {relative}")
        for candidate in path.rglob("*"):
            candidate_relative = candidate.relative_to(root)
            if "__pycache__" in candidate_relative.parts or candidate.suffix in {
                ".pyc",
                ".pyo",
            }:
                continue
            if candidate.is_symlink():
                raise ValueError(
                    "Submission archives cannot include symlinks: "
                    f"{candidate_relative}"
                )
            if candidate.is_file():
                files.append(candidate)
    return sorted(files, key=lambda path: path.relative_to(root).as_posix())


def build_archive(root: Path, output: Path) -> Path:
    manifest, _package = validate_repository(root)
    files = collect_files(root)
    output.parent.mkdir(parents=True, exist_ok=True)

    with tempfile.NamedTemporaryFile(
        dir=output.parent,
        suffix=".zip",
        delete=False,
    ) as handle:
        temporary = Path(handle.name)

    try:
        with zipfile.ZipFile(
            temporary,
            "w",
            compression=zipfile.ZIP_DEFLATED,
            compresslevel=9,
        ) as archive:
            for path in files:
                relative = path.relative_to(root).as_posix()
                mode = stat.S_IMODE(path.stat().st_mode)
                info = zipfile.ZipInfo(relative, ZIP_TIMESTAMP)
                info.create_system = 3
                info.external_attr = mode << 16
                info.compress_type = zipfile.ZIP_DEFLATED
                archive.writestr(info, path.read_bytes(), compresslevel=9)

        if temporary.stat().st_size > 100 * 1024 * 1024:
            raise ValueError("Submission ZIP exceeds the 100 MB compressed limit")
        temporary.replace(output)
    finally:
        temporary.unlink(missing_ok=True)

    digest = hashlib.sha256(output.read_bytes()).hexdigest()
    print(
        json.dumps(
            {
                "name": manifest["name"],
                "version": manifest["version"],
                "archive": str(output),
                "files": len(files),
                "bytes": output.stat().st_size,
                "sha256": digest,
            },
            indent=2,
        )
    )
    return output


def main() -> int:
    root_default = Path(__file__).resolve().parents[1]
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--root", type=Path, default=root_default)
    parser.add_argument("--output", type=Path)
    args = parser.parse_args()

    root = args.root.resolve()
    manifest = json.loads(
        (root / ".codex-plugin" / "plugin.json").read_text(encoding="utf-8")
    )
    output = args.output or root / "dist" / f"{manifest['name']}-{manifest['version']}.zip"
    build_archive(root, output.resolve())
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: scripts/build-submission-zip.mjs
#!/usr/bin/env node

import { createHash, randomUUID } from "node:crypto";
import {
  lstatSync,
  mkdirSync,
  readdirSync,
  readFileSync,
  renameSync,
  rmSync,
  writeFileSync,
} from "node:fs";
import { dirname, join, relative, resolve, sep } from "node:path";
import { fileURLToPath } from "node:url";
import { parseArgs } from "node:util";

import { createDeterministicZip } from "./lib/deterministic-zip.mjs";
import { readJson, validateRepository } from "./validate-submission.mjs";

const ROOT_DEFAULT = resolve(dirname(fileURLToPath(import.meta.url)), "..");
const INCLUDED_PATHS = [
  ".codex-plugin",
  "assets",
  "skills",
  "LICENSE",
  "README.md",
  "package.json",
];

function archivePath(root, path) {
  return relative(root, path).split(sep).join("/");
}

function collectDirectory(root, directory, files) {
  const entries = readdirSync(directory, { withFileTypes: true }).sort((left, right) =>
    left.name.localeCompare(right.name, "en"),
  );
  for (const entry of entries) {
    const path = join(directory, entry.name);
    const metadata = lstatSync(path);
    if (metadata.isSymbolicLink()) {
      throw new TypeError(
        `Submission archives cannot include symlinks: ${archivePath(root, path)}`,
      );
    }
    if (metadata.isDirectory()) {
      collectDirectory(root, path, files);
    } else if (metadata.isFile()) {
      files.push(path);
    }
  }
}

export function collectFiles(root) {
  const files = [];
  for (const item of INCLUDED_PATHS) {
    const path = join(root, item);
    let metadata;
    try {
      metadata = lstatSync(path);
    } catch {
      throw new TypeError(`Required submission path is missing: ${item}`);
    }
    if (metadata.isSymbolicLink()) {
      throw new TypeError(`Submission archives cannot include symlinks: ${item}`);
    }
    if (metadata.isFile()) {
      files.push(path);
    } else if (metadata.isDirectory()) {
      collectDirectory(root, path, files);
    } else {
      throw new TypeError(`Required submission path is not a file or directory: ${item}`);
    }
  }
  return files.sort((left, right) =>
    archivePath(root, left).localeCompare(archivePath(root, right), "en"),
  );
}

export function buildArchive(root, output) {
  const resolvedRoot = resolve(root);
  const resolvedOutput = resolve(output);
  const [manifest] = validateRepository(resolvedRoot, { reviewReady: true });
  const files = collectFiles(resolvedRoot);
  const entries = files.map((path) => {
    const metadata = lstatSync(path);
    return {
      name: archivePath(resolvedRoot, path),
      data: readFileSync(path),
      mode: metadata.mode & 0o7777,
    };
  });
  const archive = createDeterministicZip(entries);
  if (archive.length > 100 * 1024 * 1024) {
    throw new RangeError("Submission ZIP exceeds the 100 MB compressed limit");
  }

  mkdirSync(dirname(resolvedOutput), { recursive: true });
  const temporary = `${resolvedOutput}.${process.pid}.${randomUUID()}.tmp`;
  try {
    writeFileSync(temporary, archive, { mode: 0o600 });
    renameSync(temporary, resolvedOutput);
  } finally {
    rmSync(temporary, { force: true });
  }

  const digest = createHash("sha256").update(archive).digest("hex");
  console.log(
    JSON.stringify(
      {
        name: manifest.name,
        version: manifest.version,
        archive: resolvedOutput,
        files: files.length,
        bytes: archive.length,
        sha256: digest,
      },
      null,
      2,
    ),
  );
  return resolvedOutput;
}

function usage() {
  return "Usage: build-submission-zip.mjs [--root <path>] [--output <zip>]";
}

export function main(argv = process.argv.slice(2)) {
  const { values } = parseArgs({
    args: argv,
    options: {
      root: { type: "string", default: ROOT_DEFAULT },
      output: { type: "string" },
      help: { type: "boolean", short: "h" },
    },
    strict: true,
  });
  if (values.help) {
    console.log(usage());
    return 0;
  }
  const root = resolve(values.root);
  const manifest = readJson(join(root, ".codex-plugin", "plugin.json"));
  const output = values.output ?? join(root, "dist", `${manifest.name}-${manifest.version}.zip`);
  buildArchive(root, output);
  return 0;
}

if (process.argv[1] && resolve(process.argv[1]) === fileURLToPath(import.meta.url)) {
  try {
    process.exitCode = main();
  } catch (error) {
    console.error(`Submission archive failed: ${error.message}`);
    process.exitCode = 1;
  }
}
